impute_constant/impute_constant.tex

```diff
@@ -0,0 +1,196 @@
+\author{Grace Tian}
+\date{Summer 2021}
+\begin{document}
+
+
+\maketitle
+\tableofcontents
+
+\section{Algorithm Implementation}
+\subsection{Code in Rust}
+The current OpenDP library contains the \texttt{make\_impute\_constant} function implementing the impute constant function. This is defined in lines 62-75 of the file \texttt{impute.rs} in the Git repository (\url{https://github.com/opendp/opendp/blob/21-impute/rust/opendp/src/trans/impute.rs#L62-L75}).
+
+\includegraphics[width=\textwidth]{make_impute_constant.jpg}
+
+
+\subsection{Pseudo Code in Python}\label{sec:pseudocode}
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Variable \texttt{\texttt{constant}} must be of type \texttt{DA::NonNull}
+        \item Type \texttt{DA} must have traits \texttt{ImputableDomain}. % Unless the null method is defined, this means that \texttt{DA} is either \texttt{OptionNullDomain<AllDomain<T>>} or \texttt{InherentNullDomain<AllDomain<T>>}.
+        \item \texttt{DA::NonNull} must have traits \texttt{Clone}
+    \end{itemize}
+\end{itemize}
+
+\subsubsection*{Postconditions}
+
+\begin{itemize}
+    \item Either a valid \texttt{Transformation} is returned or an error is returned.
+    \item \grace{Should I also say that \texttt{impute\_constant} must satisfy the preconditions of make row by row? (See pseudo code \ref{line:row-transform}.) That is, \texttt{impute\_constant} must be a pure function?}
+\end{itemize}
+
+
+\begin{lstlisting}[language=Python,  escapechar=|]
+def make_impute_constant(constant : DA::NonNull):
+    # instead of VectorDomain(DA), we add ::new() to get a new instance of DA. This is bcause DA has the ImputableDomain trait.
+    input_domain = VectorDomain(DA::new())
+    output_domain = VectorDomain(AllDomain(DA::NonNull))
+    input_metric = SymmetricDistance()
+    output_metric = SymmetricDistance()
+    
+    assert(not constant.is_null); # not DA::is_null(constant) |\label{line:null}|
+    
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*1
+
+    def function(data: Vec[DA::Carrier]) -> Vec[DA::NonNull]: |\label{line:fn}|
+        def impute_constant(x: DA) -> DA::NonNull: |\label{line:atom-fn}|
+            return constant if x.is_null else x
+        return list(map(impute_constant, data)) |\label{line:map}|
+        
+    # can we comment out input_metric, output_metric, and stability_relation, and function if it's not being called anymore?
+    return make_row_by_row(input_domain, output_domain, impute_constant); |\label{line:row-transform}|
+\end{lstlisting}
+
+\subsection{Proof}
+
+\begin{lemma}[\texttt{DA::NonNull} contains \texttt{null}] \label{lemma:NonNull}
+A \texttt{var} of type \texttt{DA::NonNull} can be of type \texttt{null}.
+\end{lemma}
```

Comment:
Added this lemma per Mike's comments on why we must check constant for nullity in appropriate output domain proof outline.

Reply:
I don't follow what this lemma is saying.  it seems to be just a specific example (for f64), and don't want we want to say it *cannot* be of type null?

Reply:
Perhaps we should rename DA::NonNull to DA::Imputed. It represents the type after imputation. In the case of InherentNullDomain, DA::Imputed is a type like f64, which is a type that could be null. So we check that the constant is non-null at construction-time. Then in the function we check that each data element is non-null at runtime and replace with the known-to-be-non-null constant.

Reply:
Yes, I think renaming it to `Imputed` makes more sense

Reply:
It's been renamed in the repo.